### ISSUE TITLE
CSS Modules Don't Rebuild on Save

### ISSUE THREAD

**Issue by ErikAGriffin** [gatsbyjs/gatsby#34804]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

CSS/SASS Modules no longer trigger rebuilds in Gatsby.  Last version I have of this working is Gatsby 3.7.2.  
The behavior seems a bit strange, initially i was opening this bug specifically for Typescript files, but with the latest version of gatsby it seems that normal javascript files are not rebuilding when a referenced css module is changed either.

### Reproduction Link

https://github.com/ErikAGriffin/gatsby-css-modules

### Steps to Reproduce

1.  Build and run project
2.  Edit and save `src/styles/index.module.css`
3. Nothing happens
...

### Expected Result

Expected result is that Gatsby rebuilds and does Hot Module Replacement.  However it isn't just that the HMR isn't working, refreshing the page won't show the new styles either.  The change is not picked up by gatsby at all.  You must save a javascript file for Gatsby to be rebuilt and the updated css module styles to be pulled in.

### Actual Result

The change in the css module is not detected by gatsby at all.

### Environment

```shell
System:
    OS: macOS 10.15.7
    CPU: (8) x64 Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 16.6.1 - ~/.n/bin/node
    Yarn: 1.22.17 - /usr/local/bin/yarn
    npm: 7.20.3 - ~/.n/bin/npm
  Languages:
    Python: 3.9.6 - /Users/erik/.pyenv/shims/python
  Browsers:
    Chrome: 98.0.4758.80
    Safari: 14.1.1
  npmPackages:
    gatsby: ^4.7.1 => 4.7.1 
  npmGlobalPackages:
    gatsby-cli: 4.6.1
    gatsby: 3.6.1
```

**Comment by ErikAGriffin** [comment]
As part of the investigation I modified gatsby's SET_WEBPACK_CONFIG reducer to output the final webpack configuration.  The output for the css/scss module loaders looked completely fine, and identical to the output of the same loaders when this feature was working

**Comment by ErikAGriffin** [comment]
This is not stale.  Hot-reloading the page with styles from CSS modules is core functionality, and should be working in Gatsby by default.   Can I get `status:confirmed` on this?  The default Gatsby build suffers from this bug.

**Comment by cschlay** [comment]
I don't think it is stale, I had same problem and have been following this issue.

**Comment by rjp44** [comment]
I had this problem on a V4 site and found this issue (and a similar discussion #34741).

TL;DR: as far as I can tell this issue is not reproducible on a fresh install of V4 but is some sort package dependency artifact (for me at least).

Had a look at the webpack config of my broken project via a `gatsby-node.js` reducer and couldn't find anything obviously wrong (not that I'm anywhere near a webpack config expert).

Figured it must be something specific to my and a few other projects from the lack of traction these two issues have. If hot-reloading of SASS had stopped working for everyone a few versions ago then I'm sure there would be more people screaming. This is occuring for me in a project I have dragged forward from V2 to V4.

Had a look at the project packages, re-added gatsby@latest and upgraded all the deps manually. Rebuilding and hot loading of SCSS changes now works perfectly again.

I have no idea at what point the breakage occurred as I hadn't been playing with my CSS very much in develop but can confirm it now works fine for me, on a project that was broken, with the following relevant package version specs:

```
"gatsby": "^4.11.1",
"gatsby-plugin-sass": "^5.11.1",
"node-sass": "^6.0.0"
```
The main difference in the generated webpack config between broken and working was:

`"loader": "/Users/xxx/yyy/zzz/node_modules/postcss-loader/dist/cjs.js"` -> `"loader": "/Users/xxx/yyy/zzz/node_modules/gatsby/node_modules/postcss-loader/dist/cjs.js"` in a few places.

**Comment by makeable** [comment]
I have found a way to replicate this behaviour. It seems to be develop mode as a whole that has this issue.

1) install gatsby as per the docs and run develop
2) make something that will cause an error in the index page and save
3) undo the error and save

You now have broken HMR. Manual refreshing of the browser is also broken. If you close the browser tab and reopen you get the same problem.
You must close down your browser session (i.e. quit the browser) and restart for it to work again.

This is using latest Safari on latest macOS.

I have no intention of jumping through hoops to fill out that monster bug report form. If someone else wants to, feel free.

**Comment by makeable** [comment]
Maybe related: https://github.com/gatsbyjs/gatsby/issues/35563

**Comment by ErikAGriffin** [comment]
Can this be re-opened?  I'm still getting this issue on a fresh build of Gatsby:

1. Run `gatsby new`
2. Create a CSS module, import it into a page and use it to style an element
3. Edit the css-module file and save.
**Expected Behavior:**
Hot-reload triggered of site and new styles visible on page.
**Actual Behavior:**
Nothing happens

**Comment by LekoArts** [comment]
Please follow https://github.com/gatsbyjs/gatsby/issues/36225
